pkg/processor/worker/worker.go

```diff
@@ -66,6 +78,19 @@ func NewWorker(parentLogger logger.Logger,
 
 // ProcessEvent sends the event to the associated runtime
 func (w *Worker) ProcessEvent(event nuclio.Event, functionLogger logger.Logger) (interface{}, error) {
+
+	// locking worker state for the entire event processing duration to prevent concurrent draining
+	w.state.Lock()
+	defer w.state.Unlock()
+
+	// return an error if the worker is not available (e.g., terminated or drained).
+	switch w.state.State {
+	case StateDrained:
+		return nil, ErrWorkerIsDrained
+	case StateTerminated:
+		return nil, ErrWorkerIsTerminated
+	}
```

Comment:
move this block to a function `w.isAvailable`, that will do the locking and return the proper error according to the state

Reply:
@TomerShor the thing is that we block the mutex above and we cannot unlock it when state checking and event processing is happening, because otherwise it will cause sync issues.